File: csharp/WorkOOP/NGon.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WorkOOP
{
    // Реализация класса NGon – N-угольник.
    internal class NGon: IShape
    {
        protected int n;          // n – количество точек.
        protected Point2D[] p;    // p – массив двумерных точек.

        /*
         * NGon – конструктор, описывающий инициализацию объекта. На вход принимает массив точек в порядке обхода по N-угольнику.
         */
        public NGon(Point2D[] p) {
            if (p.Length == 0)
                throw new Exception("No Zero NGON!");
            this.p = p;
            n = p.Length; 
        }

        /*
         * Реализация всех функций из интерфейса IShape
         */ 
        public int getN() => n;
        public Point2D[] getP() => p;
        public void setP(Point2D[] p) {
            this.p = p; 
            n = p.Length; 
        }
        public void setP(Point2D p, int i) => this.p[i] = p;


        public double square()
        {
            if (n > 3)
            {
                Point2D[] a = p.Take(n - 1).ToArray();
                Point2D[] b = new Point2D[] { p[0], p[n - 2], p[n - 1] };
                return new NGon(a).square() + new TGon(b).square();
            }
            else 
                return new TGon(p).square();
        }

        public double length()
        {
            double l = 0;
            Point2D prevp = p[0];
            foreach (Point2D point in p) { l += new Segment(prevp, point).length(); prevp = point; }
            l += new Segment(p[0], p[n - 1]).length();
            return l;
        }
        
        public IShape shift(Point2D a) {
            for (int i = 0; i < n; i++) 
                p[i].add(a); 
            return this; 
        }

        public IShape rot(double phi) {
            for (int i = 0; i < n; i++) 
                p[i].rot(phi); 
            return this; 
        }

        public IShape symAxis(int i) {
            for (int j = 0; j < n; j++) 
                p[j].symAxis(i); 
            return this; 
        }

        public bool cross(IShape i)
        {
            Point2D prev = p[0];
            foreach (Point2D pt in p)
            {
                if (new Segment(prev, pt).cross(i))
                    return true;
                prev = pt;
            }
            if (new Segment(p[0], p[n - 1]).cross(i))
                return true;
            return false;
        }

        public String toString()
        {
            StringBuilder sb = new StringBuilder(); sb.Append("Number of headpoints: " + n + "\nPoints: [ ");
            foreach (Point2D e in p)
                sb.Append(e.toString() + " ");
            sb.Append(" ]");
            return sb.ToString();
        }
    } 
}

File: python/src/core/ngon.py
from core.IShape import IShape
from core.point2d import Point2D
from core.segment import Segment
from core.polyline import Polyline

class NGon(IShape):
	def __init__(self, p):
		self.p = p
		self.n = len(p)

	def __str__(self):
		st = f"NGon: {len(self.p)} points: "
		for i in range(len(self.p)):
			st += f"{str(self.p[i])}, "
		st = st[:len(st)-2]
		return st

	def getP(self):
		return self.p

	def setP(self, newP, i = None):
		result = NGon(self.p)
		if i == None:
			result.p = newP
		else:
			result.p[i] = newP

	def getN(self):
		return self.n

	def square(self):
		a = 0
		j = self.n - 1
		for i in range(self.n):
			a += (self.p[j].x[0] + self.p[i].x[0]) * (self.p[j].x[1] - self.p[i].x[1])
			j = i
		return abs(a/2.0)

	def length(self):
		l = 0
		pprev = self.p[0]
		for point in self.p:
			l += Segment(pprev, point).length()
			pprev = point
		l += Segment(self.p[0], self.p[self.getN() - 1]).length()
		return l

	def shift(self, point):
		result = NGon(self.p)
		for i in range(self.n):
			result.p[i] = result.p[i].add(point)
		return result

	def rot(self, phi):
		result = NGon(self.p)
		for i in range(self.n):
			result.p[i] = result.p[i].rot(phi)
		return result

	def symAxis(self, i):
		if i < 0:
			raise Exception("Ошибка: отрицательное пространство")
		if self.p[0].dim < i:
			raise Exception("Ошибка: имеет пространство меньшее, чем нужное")
		result = NGon(self.p)
		for j in range(self.n):
			result.p[i] = result.p[j].symAxis(i)
		return result

	def cross(self, figure):
		plln = Polyline(figure.n, figure.p)
		pprev = self.p[0]
		for pt in self.p[1:]:
			if Segment(pprev, pt).cross(plln):
				return True
			pprev = pt
		if Segment(self.p[0], self.p[self.n - 1]).cross(plln):
			return True
		return False